File: tools/tf-types-ts/src/core/bundle.ts
/**
 * Encrypted .tfbundle (L4/L5) sealing + transparency-log anchoring.
 *
 * sealBundle / openBundle implement the wrap-then-encrypt scheme
 * described in proof-bundle-encrypted.schema.json:
 *
 *   data_key  = random 32 bytes
 *   nonce     = random 12 bytes
 *   ciphertext = chacha20poly1305(plaintext, data_key, nonce)
 *
 *   for each recipient:
 *     ephemeral_priv = random X25519 scalar
 *     ephemeral_pub  = X25519(ephemeral_priv, base_point)
 *     shared         = X25519(ephemeral_priv, recipient_kem_public)
 *     wrap_key       = HKDF-SHA256(shared, info="tfbundle/wrap")
 *     wrap_nonce     = random 12 bytes
 *     wrapped        = chacha20poly1305(data_key, wrap_key, wrap_nonce)
 *
 * The recipient reverses the process with their X25519 secret key.
 *
 * Transparency anchoring is intentionally simple: an `Anchor` interface
 * describes how a backend submits the bundle and returns an inclusion
 * proof; v0.1.0 ships an in-process backend (good for tests) plus
 * `submitToRfc6962` and `submitToSigstore` stubs that POST a JSON body
 * the spec says is appropriate.
 *
 * RFC 3161 timestamping: `signRfc3161Timestamp` produces the canonical
 * TimeStampReq bytes (DER); a real TSA call is delegated to a callback.
 */

import { sha256 } from "@noble/hashes/sha256";
import { hkdf } from "@noble/hashes/hkdf";
import { chacha20poly1305 } from "@noble/ciphers/chacha";
import { x25519 } from "@noble/curves/ed25519";
import { canonicalize } from "./canonical.js";
import { ed25519Sign, ed25519Verify } from "./crypto.js";
import type { ProofBundle } from "../generated/proof-bundle.js";
import type { ActorId, SignatureEnvelope } from "../generated/_common.js";

export interface BundleRecipient {
  actor: ActorId;
  /** Recipient's 32-byte X25519 public key. */
  kemPublic: Uint8Array;
  /** Optional key id (recorded in the wrapped-key entry). */
  keyId?: string;
}

export interface SealBundleArgs {
  bundle: ProofBundle;
  recipients: BundleRecipient[];
  level: "L4" | "L5";
  /** Daemon's ed25519 private key for the outer signature. */
  signerPrivateKey: Uint8Array;
  /** Actor URI corresponding to signerPrivateKey. */
  signer: ActorId;
}

export interface EncryptedProofBundle {
  bundle_version: "1";
  level: "L4" | "L5";
  ciphertext: string;
  nonce: string;
  wrapped_keys: Array<{
    recipient: ActorId;
    recipient_key_id?: string;
    ephemeral_public: string;
    wrapped: string;
    wrap_nonce: string;
  }>;
  transparency_anchor?: Record<string, unknown>;
  signature: SignatureEnvelope;
}

const HKDF_INFO = new TextEncoder().encode("tfbundle/wrap");

function randBytes(len: number): Uint8Array {
  const out = new Uint8Array(len);
  crypto.getRandomValues(out);
  return out;
}

function b64(b: Uint8Array): string {
  return Buffer.from(b).toString("base64");
}

function unb64(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, "base64"));
}

export async function sealBundle(args: SealBundleArgs): Promise<EncryptedProofBundle> {
  const dataKey = randBytes(32);
  const nonce = randBytes(12);
  const plaintext = new TextEncoder().encode(canonicalize(args.bundle as unknown));
  const cipher = chacha20poly1305(dataKey, nonce);
  const ciphertext = cipher.encrypt(plaintext);

  const wrappedKeys: EncryptedProofBundle["wrapped_keys"] = [];
  for (const r of args.recipients) {
    if (r.kemPublic.length !== 32) {
      throw new Error("recipient kemPublic must be 32 bytes");
    }
    const ephemeralPriv = randBytes(32);
    const ephemeralPub = x25519.scalarMultBase(ephemeralPriv);
    const shared = x25519.scalarMult(ephemeralPriv, r.kemPublic);
    const wrapKey = hkdf(sha256, shared, undefined, HKDF_INFO, 32);
    const wrapNonce = randBytes(12);
    const wrapped = chacha20poly1305(wrapKey, wrapNonce).encrypt(dataKey);
    wrappedKeys.push({
      recipient: r.actor,
      recipient_key_id: r.keyId,
      ephemeral_public: b64(ephemeralPub),
      wrapped: b64(wrapped),
      wrap_nonce: b64(wrapNonce),
    });
  }

  const stub: EncryptedProofBundle = {
    bundle_version: "1",
    level: args.level,
    ciphertext: b64(ciphertext),
    nonce: b64(nonce),
    wrapped_keys: wrappedKeys,
    signature: { algorithm: "ed25519", signer: args.signer, signature: "" },
  };
  const digest = encryptedSigningBytes(stub);
  const sig = await ed25519Sign(digest, args.signerPrivateKey);
  stub.signature = {
    algorithm: "ed25519",
    signer: args.signer,
    signature: b64(sig),
  };
  return stub;
}

export interface OpenBundleArgs {
  encrypted: EncryptedProofBundle;
  /** Recipient's X25519 private key (32 bytes). */
  recipientPrivateKey: Uint8Array;
  /** Recipient actor URI to look up the right wrapped key. */
  recipientActor: ActorId;
  /** Optional public key for verifying the outer signature. */
  signerPublicKey?: Uint8Array;
}

export async function openBundle(args: OpenBundleArgs): Promise<ProofBundle> {
  const enc = args.encrypted;
  const wrap = enc.wrapped_keys.find((w) => w.recipient === args.recipientActor);
  if (!wrap) throw new Error(`no wrapped key for recipient ${args.recipientActor}`);
  const ephemeralPub = unb64(wrap.ephemeral_public);
  const wrapped = unb64(wrap.wrapped);
  const wrapNonce = unb64(wrap.wrap_nonce);
  const shared = x25519.scalarMult(args.recipientPrivateKey, ephemeralPub);
  const wrapKey = hkdf(sha256, shared, undefined, HKDF_INFO, 32);
  const dataKey = chacha20poly1305(wrapKey, wrapNonce).decrypt(wrapped);
  const ciphertext = unb64(enc.ciphertext);
  const nonce = unb64(enc.nonce);
  const plaintext = chacha20poly1305(dataKey, nonce).decrypt(ciphertext);
  const json = new TextDecoder().decode(plaintext);
  if (args.signerPublicKey) {
    const digest = encryptedSigningBytes(enc);
    const sig = unb64(enc.signature.signature);
    const verified = await ed25519Verify(args.signerPublicKey, digest, sig);
    if (!verified) throw new Error("encrypted bundle signature did not verify");
  }
  return JSON.parse(json) as ProofBundle;
}

export function encryptedSigningBytes(enc: EncryptedProofBundle): Uint8Array {
  const { signature: _signature, ...rest } = enc;
  void _signature;
  return sha256(new TextEncoder().encode(canonicalize(rest as unknown)));
}

/* -------------------------------------------------------------------------- */
/*  Transparency anchoring                                                    */
/* -------------------------------------------------------------------------- */

export interface AnchorBackend {
  readonly kind: "rfc6962" | "sigstore" | "custom" | "memory";
  readonly url?: string;
  submit(bundleBytes: Uint8Array): Promise<{ inclusion_proof: Record<string, unknown> }>;
  verifyInclusion(
    bundleBytes: Uint8Array,
    inclusionProof: Record<string, unknown>,
  ): Promise<boolean>;
}

/** Memory anchor — keeps a hash of every submitted bundle so callers
 *  can round-trip submit + verify in tests without external backends. */
export class MemoryAnchor implements AnchorBackend {
  readonly kind = "memory" as const;
  private entries = new Map<string, number>();
  async submit(bundleBytes: Uint8Array): Promise<{ inclusion_proof: Record<string, unknown> }> {
    const digest = Buffer.from(sha256(bundleBytes)).toString("hex");
    const seq = this.entries.size;
    this.entries.set(digest, seq);
    return { inclusion_proof: { kind: "memory", digest, sequence_number: seq } };
  }
  async verifyInclusion(
    bundleBytes: Uint8Array,
    inclusionProof: Record<string, unknown>,
  ): Promise<boolean> {
    const expected = Buffer.from(sha256(bundleBytes)).toString("hex");
    if (inclusionProof.digest !== expected) return false;
    return this.entries.get(expected) === inclusionProof.sequence_number;
  }
}

/** Submit to an RFC 6962 (Certificate Transparency) log. The backend
 *  POSTs `{ chain: [base64(bundle)] }` to `/ct/v1/add-chain` and stores
 *  the returned signed certificate timestamp (SCT) as the inclusion
 *  proof. Network errors propagate; the helper does not retry. */
export async function submitToRfc6962(
  url: string,
  bundleBytes: Uint8Array,
  fetcher: typeof fetch = fetch,
): Promise<{ inclusion_proof: Record<string, unknown> }> {
  const body = JSON.stringify({ chain: [b64(bundleBytes)] });
  const resp = await fetcher(`${url}/ct/v1/add-chain`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body,
  });
  if (!resp.ok) throw new Error(`rfc6962 anchor failed: ${resp.status} ${resp.statusText}`);
  const sct = await resp.json();
  return { inclusion_proof: { kind: "rfc6962", sct } };
}

/** Submit to a sigstore-compatible Rekor log (POST /api/v1/log/entries). */
export async function submitToSigstore(
  url: string,
  bundleBytes: Uint8Array,
  fetcher: typeof fetch = fetch,
): Promise<{ inclusion_proof: Record<string, unknown> }> {
  const body = JSON.stringify({
    apiVersion: "0.0.1",
    kind: "intoto",
    spec: { content: { hash: { algorithm: "sha256", value: Buffer.from(sha256(bundleBytes)).toString("hex") } } },
  });
  const resp = await fetcher(`${url}/api/v1/log/entries`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body,
  });
  if (!resp.ok) throw new Error(`sigstore anchor failed: ${resp.status} ${resp.statusText}`);
  const entry = await resp.json();
  return { inclusion_proof: { kind: "sigstore", entry } };
}

/* -------------------------------------------------------------------------- */
/*  RFC 3161 timestamping                                                     */
/* -------------------------------------------------------------------------- */

/** Produce the DER-encoded TimeStampReq for SHA-256 over `data`.
 *  Callers POST these bytes to a TSA (`Content-Type:
 *  application/timestamp-query`) and receive a TimeStampResp. */
export function buildRfc3161Request(data: Uint8Array): Uint8Array {
  const digest = sha256(data);
  // TimeStampReq ::= SEQUENCE {
  //   version              INTEGER  { v1(1) },
  //   messageImprint       MessageImprint,
  //   reqPolicy            OBJECT IDENTIFIER OPTIONAL,
  //   nonce                INTEGER OPTIONAL,
  //   certReq              BOOLEAN DEFAULT FALSE,
  //   extensions           [0] IMPLICIT Extensions OPTIONAL
  // }
  // MessageImprint ::= SEQUENCE { hashAlgorithm AlgorithmIdentifier, hashedMessage OCTET STRING }
  const oidSha256 = new Uint8Array([
    0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
  ]);
  const algId = derSequence([oidSha256, new Uint8Array([0x05, 0x00])]);
  const hashedMessage = derOctetString(digest);
  const messageImprint = derSequence([algId, hashedMessage]);
  const version = derInteger(new Uint8Array([0x01]));
  const certReqTrue = derBoolean(true);
  return derSequence([version, messageImprint, certReqTrue]);
}

function derSequence(parts: Uint8Array[]): Uint8Array {
  const body = concat(...parts);
  return concat(new Uint8Array([0x30]), derLen(body.length), body);
}

function derOctetString(bytes: Uint8Array): Uint8Array {
  return concat(new Uint8Array([0x04]), derLen(bytes.length), bytes);
}

function derInteger(bytes: Uint8Array): Uint8Array {
  let i = 0;
  while (i < bytes.length - 1 && bytes[i] === 0) i++;
  let payload = bytes.slice(i);
  if ((payload[0] ?? 0) & 0x80) {
    const padded = new Uint8Array(payload.length + 1);
    padded.set(payload, 1);
    payload = padded;
  }
  return concat(new Uint8Array([0x02]), derLen(payload.length), payload);
}

function derBoolean(b: boolean): Uint8Array {
  return new Uint8Array([0x01, 0x01, b ? 0xff : 0x00]);
}

function derLen(n: number): Uint8Array {
  if (n < 0x80) return new Uint8Array([n]);
  const bytes: number[] = [];
  let v = n;
  while (v > 0) {
    bytes.unshift(v & 0xff);
    v >>>= 8;
  }
  return new Uint8Array([0x80 | bytes.length, ...bytes]);
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((a, p) => a + p.length, 0);
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}

File: crates/tf-types/src/bundle.rs
//! Encrypted .tfbundle (L4/L5) sealing + transparency anchoring — Rust
//! mirror of `tools/tf-types-ts/src/core/bundle.ts`.

use base64::engine::general_purpose::STANDARD;
use base64::Engine;
use chacha20poly1305::aead::{Aead, KeyInit};
use chacha20poly1305::{ChaCha20Poly1305, Nonce};
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use hkdf::Hkdf;
use rand::RngCore;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use x25519_dalek::{PublicKey as X25519Public, StaticSecret as X25519Secret};

use crate::canonicalize;

const HKDF_INFO: &[u8] = b"tfbundle/wrap";

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct EncryptedProofBundle {
    pub bundle_version: String,
    pub level: String,
    pub ciphertext: String,
    pub nonce: String,
    pub wrapped_keys: Vec<WrappedKey>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub transparency_anchor: Option<Value>,
    pub signature: SignatureEnvelope,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct WrappedKey {
    pub recipient: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub recipient_key_id: Option<String>,
    pub ephemeral_public: String,
    pub wrapped: String,
    pub wrap_nonce: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct SignatureEnvelope {
    pub algorithm: String,
    pub signer: String,
    pub signature: String,
}

#[derive(Clone, Debug)]
pub struct BundleRecipient {
    pub actor: String,
    pub kem_public: [u8; 32],
    pub key_id: Option<String>,
}

pub fn seal_bundle(
    bundle: &Value,
    recipients: &[BundleRecipient],
    level: &str,
    signer_priv: &[u8; 32],
    signer: &str,
) -> EncryptedProofBundle {
    let mut rng = rand::thread_rng();
    let mut data_key = [0u8; 32];
    rng.fill_bytes(&mut data_key);
    let mut nonce_bytes = [0u8; 12];
    rng.fill_bytes(&mut nonce_bytes);
    let cipher = ChaCha20Poly1305::new(&data_key.into());
    let plaintext = canonicalize(bundle).unwrap_or_default();
    let ciphertext = cipher
        .encrypt(Nonce::from_slice(&nonce_bytes), plaintext.as_bytes())
        .expect("seal");

    let mut wrapped_keys = Vec::with_capacity(recipients.len());
    for r in recipients {
        let mut eph_seed = [0u8; 32];
        rng.fill_bytes(&mut eph_seed);
        let eph = X25519Secret::from(eph_seed);
        let eph_pub = X25519Public::from(&eph);
        let recipient_pub = X25519Public::from(r.kem_public);
        let shared = eph.diffie_hellman(&recipient_pub);
        let hk = Hkdf::<Sha256>::new(None, shared.as_bytes());
        let mut wrap_key = [0u8; 32];
        hk.expand(HKDF_INFO, &mut wrap_key).expect("hkdf");
        let mut wrap_nonce = [0u8; 12];
        rng.fill_bytes(&mut wrap_nonce);
        let wrap_cipher = ChaCha20Poly1305::new(&wrap_key.into());
        let wrapped = wrap_cipher
            .encrypt(Nonce::from_slice(&wrap_nonce), data_key.as_ref())
            .expect("wrap");
        wrapped_keys.push(WrappedKey {
            recipient: r.actor.clone(),
            recipient_key_id: r.key_id.clone(),
            ephemeral_public: STANDARD.encode(eph_pub.as_bytes()),
            wrapped: STANDARD.encode(&wrapped),
            wrap_nonce: STANDARD.encode(wrap_nonce),
        });
    }

    let mut stub = EncryptedProofBundle {
        bundle_version: "1".into(),
        level: level.into(),
        ciphertext: STANDARD.encode(&ciphertext),
        nonce: STANDARD.encode(nonce_bytes),
        wrapped_keys,
        transparency_anchor: None,
        signature: SignatureEnvelope {
            algorithm: "ed25519".into(),
            signer: signer.into(),
            signature: String::new(),
        },
    };
    let digest = encrypted_signing_bytes(&stub);
    let signing = SigningKey::from_bytes(signer_priv);
    let sig: Signature = signing.sign(&digest);
    stub.signature.signature = STANDARD.encode(sig.to_bytes());
    stub
}

pub fn open_bundle(
    enc: &EncryptedProofBundle,
    recipient_priv: &[u8; 32],
    recipient_actor: &str,
    signer_pub: Option<&[u8; 32]>,
) -> Result<Value, String> {
    let wrap = enc
        .wrapped_keys
        .iter()
        .find(|w| w.recipient == recipient_actor)
        .ok_or_else(|| format!("no wrapped key for recipient {}", recipient_actor))?;
    let eph_pub_bytes = STANDARD
        .decode(&wrap.ephemeral_public)
        .map_err(|e| format!("ephemeral_public base64: {}", e))?;
    let mut eph_pub_arr = [0u8; 32];
    if eph_pub_bytes.len() != 32 {
        return Err("ephemeral_public not 32 bytes".into());
    }
    eph_pub_arr.copy_from_slice(&eph_pub_bytes);
    let recipient_secret = X25519Secret::from(*recipient_priv);
    let shared = recipient_secret.diffie_hellman(&X25519Public::from(eph_pub_arr));
    let hk = Hkdf::<Sha256>::new(None, shared.as_bytes());
    let mut wrap_key = [0u8; 32];
    hk.expand(HKDF_INFO, &mut wrap_key).map_err(|e| e.to_string())?;
    let wrapped = STANDARD
        .decode(&wrap.wrapped)
        .map_err(|e| format!("wrapped base64: {}", e))?;
    let wrap_nonce = STANDARD
        .decode(&wrap.wrap_nonce)
        .map_err(|e| format!("wrap_nonce base64: {}", e))?;
    let data_key_bytes = ChaCha20Poly1305::new(&wrap_key.into())
        .decrypt(Nonce::from_slice(&wrap_nonce), wrapped.as_ref())
        .map_err(|e| format!("unwrap: {}", e))?;
    let ciphertext = STANDARD
        .decode(&enc.ciphertext)
        .map_err(|e| format!("ciphertext base64: {}", e))?;
    let nonce = STANDARD
        .decode(&enc.nonce)
        .map_err(|e| format!("nonce base64: {}", e))?;
    let mut data_key_arr = [0u8; 32];
    if data_key_bytes.len() != 32 {
        return Err("data_key not 32 bytes".into());
    }
    data_key_arr.copy_from_slice(&data_key_bytes);
    let plaintext = ChaCha20Poly1305::new(&data_key_arr.into())
        .decrypt(Nonce::from_slice(&nonce), ciphertext.as_ref())
        .map_err(|e| format!("decrypt: {}", e))?;

    if let Some(pk) = signer_pub {
        let digest = encrypted_signing_bytes(enc);
        let sig_bytes = STANDARD
            .decode(&enc.signature.signature)
            .map_err(|e| format!("signature base64: {}", e))?;
        let sig = Signature::from_slice(&sig_bytes).map_err(|e| format!("sig parse: {}", e))?;
        let vk = VerifyingKey::from_bytes(pk).map_err(|e| format!("verifying key: {}", e))?;
        if vk.verify(&digest, &sig).is_err() {
            return Err("encrypted bundle signature did not verify".into());
        }
    }

    let json: Value = serde_json::from_slice(&plaintext).map_err(|e| e.to_string())?;
    Ok(json)
}

pub fn encrypted_signing_bytes(enc: &EncryptedProofBundle) -> [u8; 32] {
    let mut value = serde_json::to_value(enc).unwrap_or(Value::Null);
    if let Value::Object(map) = &mut value {
        map.remove("signature");
    }
    let canonical = canonicalize(&value).unwrap_or_default();
    Sha256::digest(canonical.as_bytes()).into()
}

/// Build the DER-encoded TimeStampReq for SHA-256 over `data` (RFC 3161 §2.4.1).
pub fn build_rfc3161_request(data: &[u8]) -> Vec<u8> {
    let digest: [u8; 32] = Sha256::digest(data).into();
    let oid_sha256: [u8; 11] = [
        0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
    ];
    let alg_id = der_sequence(&[oid_sha256.to_vec(), vec![0x05, 0x00]]);
    let hashed_message = der_octet_string(&digest);
    let message_imprint = der_sequence(&[alg_id, hashed_message]);
    let version = der_integer(&[0x01]);
    let cert_req = vec![0x01, 0x01, 0xff];
    der_sequence(&[version, message_imprint, cert_req])
}

fn der_sequence(parts: &[Vec<u8>]) -> Vec<u8> {
    let body: Vec<u8> = parts.iter().flat_map(|p| p.clone()).collect();
    let mut out = Vec::with_capacity(2 + body.len());
    out.push(0x30);
    out.extend_from_slice(&der_len(body.len()));
    out.extend_from_slice(&body);
    out
}

fn der_octet_string(bytes: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(2 + bytes.len());
    out.push(0x04);
    out.extend_from_slice(&der_len(bytes.len()));
    out.extend_from_slice(bytes);
    out
}

fn der_integer(bytes: &[u8]) -> Vec<u8> {
    let mut start = 0usize;
    while start < bytes.len() - 1 && bytes[start] == 0 {
        start += 1;
    }
    let payload = &bytes[start..];
    let needs_pad = payload[0] & 0x80 != 0;
    let len = payload.len() + if needs_pad { 1 } else { 0 };
    let mut out = Vec::with_capacity(2 + len);
    out.push(0x02);
    out.extend_from_slice(&der_len(len));
    if needs_pad {
        out.push(0x00);
    }
    out.extend_from_slice(payload);
    out
}

fn der_len(n: usize) -> Vec<u8> {
    if n < 0x80 {
        return vec![n as u8];
    }
    let mut bytes = Vec::new();
    let mut v = n;
    while v > 0 {
        bytes.insert(0, (v & 0xff) as u8);
        v >>= 8;
    }
    let mut out = Vec::with_capacity(1 + bytes.len());
    out.push(0x80 | bytes.len() as u8);
    out.extend_from_slice(&bytes);
    out
}

/// In-memory transparency anchor for tests.
#[derive(Default)]
pub struct MemoryAnchor {
    entries: std::sync::Mutex<std::collections::HashMap<String, usize>>,
}

impl MemoryAnchor {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn submit(&self, bundle_bytes: &[u8]) -> Value {
        let digest = Sha256::digest(bundle_bytes);
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        let mut entries = self.entries.lock().unwrap();
        let seq = entries.len();
        entries.insert(hex.clone(), seq);
        serde_json::json!({ "kind": "memory", "digest": hex, "sequence_number": seq })
    }

    pub fn verify_inclusion(&self, bundle_bytes: &[u8], inclusion_proof: &Value) -> bool {
        let digest = Sha256::digest(bundle_bytes);
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        if inclusion_proof.get("digest").and_then(|v| v.as_str()) != Some(&hex) {
            return false;
        }
        let seq = inclusion_proof
            .get("sequence_number")
            .and_then(|v| v.as_u64())
            .map(|n| n as usize);
        let entries = self.entries.lock().unwrap();
        seq == entries.get(&hex).copied()
    }
}
